File: apps/api/app/api/routes/masterdata.py
"""Master data endpoints - clients, products, branches, etc.

These endpoints provide reference data for order management UI.
Optimized for fast loading with minimal data transfer.
"""

import logging
from fastapi import APIRouter, BackgroundTasks

from ...core.supabase import get_supabase_client
from ...services.rag_sync import sync_client_to_rag

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/masterdata", tags=["masterdata"])


@router.get("/clients")
async def get_clients():
    """Get all clients ordered by name."""
    logger.info("Fetching clients")
    supabase = get_supabase_client()

    try:
        result = (
            supabase.table("clients")
            .select("*")
            .order("name", desc=False)
            .execute()
        )
        return {"clients": result.data or []}
    except Exception as e:
        logger.error(f"Error fetching clients: {e}")
        return {"clients": []}


@router.get("/products")
async def get_products(active_only: bool = True, category: str = None):
    """Get products with optional filters."""
    logger.info(f"Fetching products: active_only={active_only}, category={category}")
    supabase = get_supabase_client()

    try:
        query = supabase.table("products").select("*")

        if active_only:
            query = query.eq("is_active", True)

        if category:
            # Support comma-separated categories like "PT,PP"
            categories = [c.strip() for c in category.split(",")]
            query = query.in_("category", categories)

        result = query.order("name", desc=False).execute()
        return {"products": result.data or []}
    except Exception as e:
        logger.error(f"Error fetching products: {e}")
        return {"products": []}


@router.get("/branches")
async def get_branches():
    """Get all branches with client info."""
    logger.info("Fetching branches")
    supabase = get_supabase_client()

    try:
        result = (
            supabase.table("branches")
            .select("*, client:clients(id, name)")
            .order("created_at", desc=True)
            .execute()
        )
        return {"branches": result.data or []}
    except Exception as e:
        logger.error(f"Error fetching branches: {e}")
        return {"branches": []}


@router.get("/receiving-schedules")
async def get_receiving_schedules():
    """Get receiving schedules ordered by day and time."""
    logger.info("Fetching receiving schedules")
    supabase = get_supabase_client()

    try:
        result = (
            supabase.table("receiving_schedules")
            .select("*")
            .order("day_of_week", desc=False)
            .order("start_time", desc=False)
            .execute()
        )
        return {"schedules": result.data or []}
    except Exception as e:
        logger.error(f"Error fetching receiving schedules: {e}")
        return {"schedules": []}


@router.get("/product-configs")
async def get_product_configs():
    """Get product configurations with product details."""
    logger.info("Fetching product configs")
    supabase = get_supabase_client()

    try:
        result = (
            supabase.table("product_config")
            .select("*, product:products!product_config_product_id_fkey(id, name, description, weight, price)")
            .order("created_at", desc=True)
            .execute()
        )
        return {"configs": result.data or []}
    except Exception as e:
        logger.error(f"Error fetching product configs: {e}")
        return {"configs": []}


@router.get("/client-frequencies")
async def get_client_frequencies():
    """Get active client delivery frequencies."""
    logger.info("Fetching client frequencies")
    supabase = get_supabase_client()

    try:
        result = (
            supabase.table("client_frequencies")
            .select("*")
            .eq("is_active", True)
            .execute()
        )
        return {"frequencies": result.data or []}
    except Exception as e:
        logger.error(f"Error fetching client frequencies: {e}")
        return {"frequencies": []}


@router.get("/vehicles")
async def get_vehicles():
    """Get all vehicles."""
    logger.info("Fetching vehicles")
    supabase = get_supabase_client()

    try:
        result = (
            supabase.table("vehicles")
            .select("*")
            .order("vehicle_code", desc=False)
            .execute()
        )
        return {"vehicles": result.data or []}
    except Exception as e:
        logger.error(f"Error fetching vehicles: {e}")
        return {"vehicles": []}


@router.get("/drivers")
async def get_drivers():
    """Get all users with driver role."""
    logger.info("Fetching drivers")
    supabase = get_supabase_client()

    try:
        # Get users with driver role and active status
        result = (
            supabase.table("users")
            .select("id, name, email, cedula")
            .eq("role", "driver")
            .eq("status", "active")
            .order("name", desc=False)
            .execute()
        )
        logger.info(f"Found {len(result.data or [])} drivers")
        return {"drivers": result.data or []}
    except Exception as e:
        logger.error(f"Error fetching drivers: {e}")
        return {"drivers": []}


@router.post("/clients/{client_id}/sync-rag")
async def sync_client_rag(client_id: str, background_tasks: BackgroundTasks):
    """Sync a client to the vector search table. Call after create/update."""
    background_tasks.add_task(sync_client_to_rag, client_id)
    return {"status": "queued", "client_id": client_id}


@router.post("/clients/sync-rag-all")
async def sync_all_clients_rag():
    """Sync ALL clients to the vector search table."""
    logger.info("Syncing all clients to RAG")
    supabase = get_supabase_client()

    try:
        result = supabase.table("clients").select("id").execute()
        clients = result.data or []

        results = []
        for client in clients:
            try:
                r = await sync_client_to_rag(client["id"])
                results.append(r)
            except Exception as e:
                logger.error(f"Failed to sync client {client['id']}: {e}")
                results.append({"status": "error", "client_id": client["id"], "error": str(e)})

        synced = sum(1 for r in results if r.get("status") == "synced")
        errors = sum(1 for r in results if r.get("status") == "error")

        return {
            "status": "completed",
            "total": len(clients),
            "synced": synced,
            "errors": errors,
        }
    except Exception as e:
        logger.error(f"Error syncing all clients: {e}")
        return {"status": "error", "message": str(e)}

File: apps/web/lib/api/masterdata.ts
"use server"

/**
 * Master Data Server Actions
 *
 * Shared across all V2 modules - reusable anywhere in the app.
 * All data fetched through FastAPI backend (no direct Supabase calls).
 */

const API_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000"

// === Types ===

export interface Client {
  id: string
  name: string
  email: string | null
  phone: string | null
  address: string | null
  nit: string | null
  is_active: boolean
  created_at: string
}

export interface Product {
  id: string
  name: string
  description: string | null
  price: number | null
  weight: number | null
  category: string | null
  is_active: boolean
  created_at: string
}

export interface Branch {
  id: string
  name: string
  address: string | null
  client_id: string
  client: { id: string; name: string } | null
  created_at: string
}

export interface ReceivingSchedule {
  id: string
  branch_id: string
  day_of_week: number
  start_time: string
  end_time: string
  created_at: string
}

export interface ProductConfig {
  id: string
  product_id: string
  product: { id: string; name: string; description: string | null; weight: number | null; price: number | null } | null
  config_type: string
  config_value: string | null
  created_at: string
}

export interface ClientFrequency {
  id: string
  client_id: string
  branch_id: string
  day_of_week: number
  is_active: boolean
  created_at: string
}

// === API Response Type ===

type ApiResult<T> = { data: T | null; error: string | null }

// === Server Actions ===

export async function getClients(): Promise<ApiResult<Client[]>> {
  try {
    const response = await fetch(`${API_URL}/api/masterdata/clients`, { cache: "no-store" })
    if (!response.ok) {
      const error = await response.json().catch(() => ({ detail: "Unknown error" }))
      return { data: null, error: error.detail || `Error: ${response.status}` }
    }
    const data = await response.json()
    return { data: data.clients, error: null }
  } catch (err) {
    return { data: null, error: err instanceof Error ? err.message : "Error fetching clients" }
  }
}

export async function getProducts(params?: {
  activeOnly?: boolean
  category?: string
}): Promise<ApiResult<Product[]>> {
  try {
    const searchParams = new URLSearchParams()
    if (params?.activeOnly !== undefined) searchParams.set("active_only", params.activeOnly.toString())
    if (params?.category) searchParams.set("category", params.category)

    const query = searchParams.toString()
    const url = `${API_URL}/api/masterdata/products${query ? `?${query}` : ""}`

    const response = await fetch(url, { cache: "no-store" })
    if (!response.ok) {
      const error = await response.json().catch(() => ({ detail: "Unknown error" }))
      return { data: null, error: error.detail || `Error: ${response.status}` }
    }
    const data = await response.json()
    return { data: data.products, error: null }
  } catch (err) {
    return { data: null, error: err instanceof Error ? err.message : "Error fetching products" }
  }
}

export async function getBranches(): Promise<ApiResult<Branch[]>> {
  try {
    const response = await fetch(`${API_URL}/api/masterdata/branches`, { cache: "no-store" })
    if (!response.ok) {
      const error = await response.json().catch(() => ({ detail: "Unknown error" }))
      return { data: null, error: error.detail || `Error: ${response.status}` }
    }
    const data = await response.json()
    return { data: data.branches, error: null }
  } catch (err) {
    return { data: null, error: err instanceof Error ? err.message : "Error fetching branches" }
  }
}

export async function getReceivingSchedules(): Promise<ApiResult<ReceivingSchedule[]>> {
  try {
    const response = await fetch(`${API_URL}/api/masterdata/receiving-schedules`, { cache: "no-store" })
    if (!response.ok) {
      const error = await response.json().catch(() => ({ detail: "Unknown error" }))
      return { data: null, error: error.detail || `Error: ${response.status}` }
    }
    const data = await response.json()
    return { data: data.schedules, error: null }
  } catch (err) {
    return { data: null, error: err instanceof Error ? err.message : "Error fetching receiving schedules" }
  }
}

export async function getProductConfigs(): Promise<ApiResult<ProductConfig[]>> {
  try {
    const response = await fetch(`${API_URL}/api/masterdata/product-configs`, { cache: "no-store" })
    if (!response.ok) {
      const error = await response.json().catch(() => ({ detail: "Unknown error" }))
      return { data: null, error: error.detail || `Error: ${response.status}` }
    }
    const data = await response.json()
    return { data: data.configs, error: null }
  } catch (err) {
    return { data: null, error: err instanceof Error ? err.message : "Error fetching product configs" }
  }
}

export async function getClientFrequencies(): Promise<ApiResult<ClientFrequency[]>> {
  try {
    const response = await fetch(`${API_URL}/api/masterdata/client-frequencies`, { cache: "no-store" })
    if (!response.ok) {
      const error = await response.json().catch(() => ({ detail: "Unknown error" }))
      return { data: null, error: error.detail || `Error: ${response.status}` }
    }
    const data = await response.json()
    return { data: data.frequencies, error: null }
  } catch (err) {
    return { data: null, error: err instanceof Error ? err.message : "Error fetching client frequencies" }
  }
}

// === Helper Functions ===

/**
 * Get branches filtered by client ID
 */
export async function getBranchesByClient(clientId: string): Promise<ApiResult<Branch[]>> {
  const result = await getBranches()
  if (result.error || !result.data) return result

  const filtered = result.data.filter(b => b.client_id === clientId)
  return { data: filtered, error: null }
}

/**
 * Get frequencies filtered by branch ID
 */
export async function getFrequenciesByBranch(branchId: string): Promise<ApiResult<ClientFrequency[]>> {
  const result = await getClientFrequencies()
  if (result.error || !result.data) return result

  const filtered = result.data.filter(f => f.branch_id === branchId)
  return { data: filtered, error: null }
}

/**
 * Get schedules filtered by branch ID
 */
export async function getSchedulesByBranch(branchId: string): Promise<ApiResult<ReceivingSchedule[]>> {
  const result = await getReceivingSchedules()
  if (result.error || !result.data) return result

  const filtered = result.data.filter(s => s.branch_id === branchId)
  return { data: filtered, error: null }
}

/**
 * Get only finished products (categories PT, PP)
 */
export async function getFinishedProducts(): Promise<ApiResult<Product[]>> {
  return getProducts({ activeOnly: true, category: "PT,PP" })
}
